src/v/cloud_topics/core/tests/event_filter_test.cc

```diff
@@ -0,0 +1,183 @@
+        req.set_value(ss::circular_buffer<model::record_batch>());
+    }
+    std::ignore = co_await std::move(write);
+    co_return;
+}
+
+TEST_CORO(EventFilterTest, filter_has_memory) {
+    // Check that the ordering between the write request submission
+    // and filter subscription doesn't matter. Previous test checks
+    // a situation when the subscription is created first and then
+    // the write request is submitted. This test checks the opposite.
+    // The write request is submitted and then the write request is
+    // submitted.
+    model::test::record_batch_spec spec{
+      .offset = model::offset{0},
+      .allow_compression = false,
+      .count = 100,
+      .records = 10,
+    };
+    auto batches = co_await model::test::make_random_batches(spec);
+    size_t reader_size_bytes = get_serialized_size(batches);
+    auto reader = model::make_memory_record_batch_reader(std::move(batches));
+    core::write_pipeline<ss::lowres_clock> pipeline;
+    auto stage = pipeline.register_pipeline_stage();
+    auto write = pipeline.write_and_debounce(
+      model::controller_ntp, std::move(reader), 1s);
+    // wait until submitted
+    // this is needed because 'write_and_debounce' has a scheduling point
+    // before the write request is added to the list
+    co_await do_until([&pipeline] {
+        core::write_pipeline_accessor accessor{.pipeline = &pipeline};
+        return accessor.write_requests_pending(1);
+    });
+    core::event_filter<ss::lowres_clock> flt(
+      core::event_type::new_write_request, stage);
+    auto event = co_await pipeline.subscribe(flt);
+    ASSERT_EQ_CORO(event.pending_write_bytes, reader_size_bytes);
+    auto pending = pipeline.get_write_requests(
+      std::numeric_limits<size_t>::max(), stage);
+    for (auto& req : pending.ready) {
+        req.set_value(ss::circular_buffer<model::record_batch>());
+    }
+    std::ignore = co_await std::move(write);
+    co_return;
+}
+
+TEST_CORO(EventFilterTest, filter_shutdown) {
+    core::write_pipeline<ss::lowres_clock> pipeline;
+    auto stage = pipeline.register_pipeline_stage();
+    // The subscription mechanism uses external abort source
+    core::event_filter<ss::lowres_clock> flt(
+      core::event_type::new_write_request, stage);
+    ss::abort_source as;
+    auto sub = pipeline.subscribe(flt, as);
+    co_await ss::sleep(10ms);
+    as.request_abort();
+    auto res = co_await std::move(sub);
+    ASSERT_TRUE_CORO(res.type == core::event_type::shutting_down);
+    co_return;
+}
+
+TEST_CORO(EventFilterTest, filter_timedout) {
+    core::write_pipeline<ss::lowres_clock> pipeline;
+    auto stage = pipeline.register_pipeline_stage();
+    // The subscription mechanism can be aborted by timeout
+    core::event_filter<ss::lowres_clock> flt(
+      core::event_type::new_write_request,
+      stage,
+      ss::lowres_clock::now() + 1ms);
+    auto sub = pipeline.subscribe(flt);
+    co_await ss::sleep(10ms);
```

Comment:
Why you are not using manual clocks here?

Reply:
why not?